pkg/build/builder/common_test.go

```diff
@@ -197,6 +197,19 @@ func Test_addBuildParameters(t *testing.T) {
 			original: `# no FROM instruction`,
 			want:     want{},
 		},
+		{
+			original: heredoc.Doc(`
+				ARG GOLANG_CONTAINER=golang:latest
+				FROM $GOLANG_CONTAINER
+				RUN echo "hello world"
+				`),
+			want: want{
+				Out: heredoc.Doc(`
+				FROM $GOLANG_CONTAINER
+				RUN echo "hello world"
+				`),
+			},
```

Comment:
1) why are we expecting this to strip the `ARG` directive from the dockerfile?

2) don't we need a test that proves we actually substitute the FROM line correctly? e.g.

```
FROM centos
ARG foo=bar
FROM $foo
```
in a buildconfig that has a dockerstrategy.From section?

Reply:
These unit tests only deal with the imagebuilder translations prior to passing the dockerfile to buildah

If we expect buildah to handle the case above on its own, we would not add a test here.

Ultimately, when this is fixed, we'll need an builds e2e in openshift/origin like the ones you linked to yesterday that do a dockerfile like the one you've noted.

Reply:
The unit test I added caused the panic before my fix.

It does not panic after my fix.

Reply:
> These unit tests only deal with the imagebuilder translations prior to passing the dockerfile to buildah

ok, but why do we expect that translation to strip the ARG directive?


> If we expect buildah to handle the case above on its own, we would not add a test here.

not the same thing.  There are two ways for a FROM reference in a dockerfile to be replaced:

1) something in your buildconfig which tells us to replace specific FROM references with a resolved imagestream.  this has nothing to do w/ buildah and nothing to do with any ARGs

2) using an ARG reference in a FROM line in the dockerfile.  This is the substitution buildah should be doing (but is not handling correctly today in all cases)

I'm referring to case (1), not case (2).

That said, upon further reflection i don't know that we *need* to add (1) per se.  It might be nice to have (to ensure we can do a proper substitution even when the FROM value is "$foo" and not a valid pullspec) but it's not a bug/regression that is being fixed in this PR.

> Ultimately, when this is fixed, we'll need an builds e2e in openshift/origin like the ones you linked to yesterday that do a dockerfile like the one you've noted.

No, the e2e we'll need at that point would have to be one where we do *not* have a dockerstrategy From imagestream reference, so that the dockerfile FROM $foo gets substituted by buildah via the ARG reference, not by the openshift "replace the last FROM with the imagestream reference" logic.

Reply:
I think the ARG is getting stripped in `ManageDockerfile()`.  The `replaceImagesFromSource()` function calls `imagebuilder.NewStages()`, which modifies the parsed tree(!), discarding any instructions that precede the first FROM instruction while doing so, so the reconstructed Dockerfile that it writes doesn't get to include them.

I'm working on a patch that pulls instructions from the parsed tree beforehand and then re-prepends them to the written Dockerfile to work around this.  It doesn't solve the substitutions-aren't-happening part of things, because we don't call into any imagebuilder functions that perform that resolution.